src/design_notes/general_coroutines.md

```diff
@@ -292,12 +293,17 @@ fn merge_gen_state<T>(f: impl FnMut() -> GeneratorState<T, T>) -> T { ... }
 ## Movability
 
 - All proposals want movability/`impl Unpin` to be inferred.
-  - Exact inference rules may only be revealed by an attempt at implementation.
+  - If we forbid "borrowed data escaping into closure state", the inference
+    rules should be relatively simple: witnessing any borrow triggers
+    immovability.
```

Comment:
This doesn't really work. For example:

```
|x: &Foo| {
    let y = &x.field;
    yield;
    drop(y);
}
```

Here `y` is live across a yield but it does not require pinning. 
```

Reply:
For that example to compile, we would need to allow borrowed inputs to escape into closure state. In that case, the inference is significantly more complex. But otherwise you'd get an error on the yield, something like:
```
3 |    yield;
  |    ---- borrowed input `x` escapes into closure state
4 |    drop(y);
  |         - borrow must enter state because it is used here
```
and the inference would never have to figure out that `y` is *not* borrowing from elsewhere in the witness.

Obviously this error isn't the best. But since closures similarly forbid moving borrowed inputs into capture state, it is the easiest thing to do for the time being.

Reply:
@samsartor hmm, so, yes. I agree that if `&Foo` here means "each time the closure is invoked, it will be supplied a fresh reference", then I guess we really can't keep the reference live across the yield. There's definitely some subtlety here. Anyway, I'm going to merge this PR, and thank you again for the effort to create it!